### PR TITLE
Added unit testing

### PR DESCRIPTION
**Pull request by freave-public**
Changes proposed in this pull request:
* This merge request adds unit testing the CLI using the Mocha testing framework.
* Removed installers from the readme in favor of installing the CLI through NPM.
* Some minor refactoring of the services.
* Ignored `.DS_STORE` (MacOS)

### CODE REVIEW

**Comment by TLadd** [comment]
Ran into some errors when running the tests:

```1) service http
       should perform a http post call:

      AssertionError [ERR_ASSERTION]: Mocks not yet satisfied:
POST https://gateway.stackpath.com:443/cdn/v1/stacks/1/sites/2/scripts
      + expected - actual

      -false
      +true

      at Scope.done (node_modules/nock/lib/scope.js:155:10)
      at Object.finally (node_modules/fancy-test/lib/nock.js:30:29)
      at Context.run (node_modules/fancy-test/lib/base.js:59:36)

  2) service http
       should perform a http patch call:

      AssertionError [ERR_ASSERTION]: Mocks not yet satisfied:
PATCH https://gateway.stackpath.com:443/cdn/v1/stacks/1/sites/2/scripts/3
      + expected - actual

      -false
      +true

      at Scope.done (node_modules/nock/lib/scope.js:155:10)
      at Object.finally (node_modules/fancy-test/lib/nock.js:30:29)
      at Context.run (node_modules/fancy-test/lib/base.js:59:36)

  3) service http
       should throw a http error:

      AssertionError [ERR_ASSERTION]: Mocks not yet satisfied:
POST https://gateway.stackpath.com:443/error
      + expected - actual

      -false
      +true

      at Scope.done (node_modules/nock/lib/scope.js:155:10)
      at Object.finally (node_modules/fancy-test/lib/nock.js:30:29)
      at Context.run (node_modules/fancy-test/lib/base.js:59:36)

  4) service http
       should not throw a http error:
     AssertionError: expected [Error: An error occurred when connecting to StackPath host https://gateway.stackpath.com.] to not exist
      at Context.test_1.test.nock.it (test/services/http.test.ts:36:25)```

File: test/services/validation.test.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+const deepCloneArray = (array: any[]) => array.map(x => Object.assign({}, x));
+
+describe("Validation, validateConfiguration", () => {
+  it("Should succeed when the configuration is valid", () => {
+    expect(() => validateConfiguration(validTestConfig)).to.not.throw();
+  });
+
+  it("Should error if stack_id is missing", () => {
+    const invalidConfig = { ...validTestConfig, stack_id: undefined };
```
Comment (TLadd) [inline test/services/validation.test.ts]: I would do
```
const invalidConfig = { ...validTestConfig, stack_id: undefined };
```

to allow the invalidConfig to be passed into validateConfiguration but not removing all other type checking on the line.

File: test/services/http.test.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      }
+    });
+
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockHttpError(false))
+    .it("should not throw a http error", async () => {
+      try {
+        await request("POST", `/error`, {}, false);
+      } catch (e) {
+        expect(e).to.not.exist;
```
Comment (TLadd) [inline test/services/http.test.ts]: This test isn't working as intended. The catch won't be triggered and so no assertions are actually running. I would change to
```
      let error;
      try {
        await request("POST", `/error`, {});
      } catch (e) {
        error = e;
      }
      expect(error).to.not.exist;
```

File: test/services/http.test.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockDeployPatch(false))
+    .it("should perform a http patch call", async () => {
+      await request("PATCH", `/cdn/v1/stacks/1/sites/2/scripts/3`);
+    });
+
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockHttpError(false))
+    .it("should throw a http error", async () => {
```
Comment (TLadd) [inline test/services/http.test.ts]: I removed the error behavior of the http service in my last PR, so this test is no longer accurate.

File: test/services/http.test.ts (hunk 2 above)
Comment (TLadd) [inline test/services/http.test.ts]: .to.not.exist; is triggering a tslint error right now too. We should pull in https://www.npmjs.com/package/tslint-no-unused-expression-chai and extend it in our tsling.json.

File: test/helpers/mock-helper.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+import {
+  STACKPATH_CREDENTIALSFILE_PATH,
+  STACKPATH_CONFIGFILE_PATH
+} from "../../src/api/constants";
+
+/*
+ * Namespace for mocking in the test environment.
+ */
+export namespace MockHelper {
```
Comment (TLadd) [inline test/helpers/mock-helper.ts]: We should remove the namespace and just import by `import * as MockHelper`

File: test/services/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
+// tslint:disable newline-per-chained-call no-implicit-dependencies
```
Comment (TLadd) [inline test/services/validation.test.ts]: Don't believe this ignore is needed.

File: test/helpers/mock-helper.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+      .withArgs(`${process.cwd()}/edgeengine/new_test.js`)
+      .callsFake(() => {
+        return true;
+      });
+    (fs.existsSync as sinon.SinonStub).callThrough();
+  };
+
+  export const restoreFileSystem = () => {
+    if (typeof (fs.writeFileSync as any)["restore"] === "function") {
+      (fs.writeFileSync as sinon.SinonStub)["restore"]();
```
Comment (TLadd) [inline test/helpers/mock-helper.ts]: Lint error here.

File: test/services/validation.test.ts (hunk 1 above)
Comment (freave-public) [reply test/services/validation.test.ts]: Thanks for your feedback! I don't see a difference between my code and your suggestion. Could you elaborate on that?

File: test/helpers/mock-helper.ts (hunk 4 above)
Comment (freave-public) [reply test/helpers/mock-helper.ts]: I think combining the mocking functions into a namespace is for typing much better. Why would we remove this namespace? You can interpret the MockHelper as an internal module (see https://www.typescriptlang.org/docs/handbook/namespaces.html as well).

Comment (TLadd) [reply test/helpers/mock-helper.ts]: Namespaces are considered a proprietary, legacy feature at this point, replaced by ES Modules. This came up a lot when Typescript was being added to babel https://github.com/babel/babel/issues/8244#issuecomment-401616905. tslint:recommended has a rule no-namespaces for this reason. Removing the `export namespace MockHelper` block from this file, resulting in the functions being named exports and using `import * as MockHelper` in the test files result in the same object, so there shouldn't be any difference from a typing perspective.

File: test/services/validation.test.ts (hunk 1 above)
Comment (TLadd) [reply test/services/validation.test.ts]: Sorry, added an `: any` to the line

File: test/services/http.test.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockHttpError(false))
+    .it("should not throw a http error", async () => {
+      let error = {};
+      try {
+        await request("POST", `/error`, {});
+      } catch (e) {
+        error = e;
+      }
+      expect(error).to.be.a("Error");
```
Comment (TLadd) [inline test/services/http.test.ts]: Should be asserting that there is not an error.

```
      let error: Error | undefined;
      try {
        await request("POST", `/error`, {});
      } catch (e) {
        error = e;
      }
      expect(error).to.be.undefined;
```

File: test/services/http.test.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockDeployPatch(false))
+    .it("should perform a http patch call", async () => {
+      await request("PATCH", `/cdn/v1/stacks/1/sites/2/scripts/3`);
+    });
+
+  test
+    .nock(STACKPATH_HOST, MockHelper.mockHttpError(false))
+    .it("should not throw a http error", async () => {
+      let error = {};
```
Comment (TLadd) [inline test/services/http.test.ts]:
```suggestion
      let error: Error | undefined;
```

Test fails still currently because `{}` is not undefined.

**Pull request merged**